src/query-builder/merge-query-builder.ts

```diff
@@ -0,0 +1,906 @@
+    K1 extends JoinReferenceExpression<DB, TT, TE>,
+    K2 extends JoinReferenceExpression<DB, TT, TE>
+  >(
+    sourceTable: TE,
+    k1: K1,
+    k2: K2
+  ): ExtractWheneableMergeQueryBuilder<DB, TT, TE, O>
+
+  using<
+    TE extends TableExpression<DB, TT>,
+    FN extends JoinCallbackExpression<DB, TT, TE>
+  >(
+    sourceTable: TE,
+    callback: FN
+  ): ExtractWheneableMergeQueryBuilder<DB, TT, TE, O>
+
+  using(...args: any): any {
+    return new WheneableMergeQueryBuilder({
+      ...this.#props,
+      queryNode: MergeQueryNode.cloneWithUsing(
+        this.#props.queryNode,
+        parseJoin('Using', args)
+      ),
+    })
+  }
+}
+
+preventAwait(
+  MergeQueryBuilder,
+  "don't await MergeQueryBuilder instances directly. To execute the query you need to call `execute` when available."
+)
+
+export interface MergeQueryBuilderProps {
+  readonly queryId: QueryId
+  readonly queryNode: MergeQueryNode
+  readonly executor: QueryExecutor
+}
+
+export class WheneableMergeQueryBuilder<
+  DB,
+  TT extends keyof DB,
+  ST extends keyof DB,
+  O
+> implements Compilable<O>, OperationNodeSource
+{
+  readonly #props: MergeQueryBuilderProps
+
+  constructor(props: MergeQueryBuilderProps) {
+    this.#props = freeze(props)
+  }
+
+  /**
+   * Adds a simple `when matched` clause to the query.
+   *
+   * For a `when matched` clause with an `and` condition, see {@link whenMatchedAnd}.
+   *
+   * For a simple `when not matched` clause, see {@link whenNotMatched}.
+   *
+   * For a `when not matched` clause with an `and` condition, see {@link whenNotMatchedAnd}.
+   *
+   * ### Examples
+   *
+   * ```ts
+   * const result = await db.mergeInto('person')
+   *   .using('pet', 'person.id', 'pet.owner_id')
+   *   .whenMatched()
+   *   .thenDelete()
+   *   .execute()
+   * ```
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * merge into "person"
+   * using "pet" on "person"."id" = "pet"."owner_id"
+   * when matched then
+   *   delete
+   * ```
+   */
+  whenMatched(): MatchedThenableMergeQueryBuilder<DB, TT, ST, TT | ST, O> {
+    return this.#whenMatched([])
+  }
+
+  /**
+   * Adds the `when matched` clause to the query with an `and` condition.
+   *
+   * This method is similar to {@link SelectQueryBuilder.where}, so see the documentation
+   * for that method for more examples.
+   *
+   * For a simple `when matched` clause (without an `and` condition) see {@link whenMatched}.
+   *
+   * ### Examples
+   *
+   * ```ts
+   * const result = await db.mergeInto('person')
+   *   .using('pet', 'person.id', 'pet.owner_id')
+   *   .whenMatchedAnd('person.first_name', '=', 'John')
+   *   .thenDelete()
+   *   .execute()
+   * ```
+   *
+   * The generated SQL (PostgreSQL):
+   *
+   * ```sql
+   * merge into "person"
+   * using "pet" on "person"."id" = "pet"."owner_id"
+   * when matched and "person"."first_name" = $1 then
+   *   delete
+   * ```
+   */
+  whenMatchedAnd<RE extends ReferenceExpression<DB, TT | ST>>(
```

Comment:
I found it did great things to typescript performance to hoist all complex argument types (that would create a lot of new type instantiations) to generics. My theory (pulled straight from my butt) is that typescript doesn't instantiate and compare the argument types when they are generics when comparing the parent interface.

So here you'd use:

```ts
whenMatchedAnd<
  RE extends ReferenceExpression<DB, TT | ST>,
  VE extends OperandValueExpressionOrList<DB, TT | ST, RE>
>(
  lhs: RE,
  op: ComparisonOperatorExpression,
  rhs: VE
): MatchedThenableMergeQueryBuilder<DB, TT, ST, TT | ST, O>
```

All `where` and other methods of this family now use this style, so maybe you could switch to that here even if there's no performance benefit.

Reply:
user time went down by 1 sec. 👌